Ref count search response bytes

original-brownbear opened the pull request:
Final step in  #102030 ... actually makes `SearchHit` read a releasable bytes reference.
Does still fallback to copying to unrolled buffers here and there which can be removed in follow-ups where it's worth the effort (aggs being the most important one probably).

Hard to create very reliable benchmarks for this because all our macro-benchmarks are quite noisy. Running http logs and PMC though, there's a statistically significant reduction in GC and reduced tail latencies in most benchmarks.

The overhead for ref-counting these bytes isn't visible in profiling as far as I can tell and for large source values, no corresponding large `byte[]` are created any longer outside of the few remaining spots where we copy to pooled buffers.

closes #102657

original-brownbear committed:
Fix more leaked SearchResponse instances in tests

original-brownbear committed:
save

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
clippyh

original-brownbear committed:
fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-itself' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix ml

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix eql

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-itself' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fixe

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix

original-brownbear committed:
ix

original-brownbear committed:
ix

original-brownbear committed:
retain ref async

original-brownbear committed:
Fix SearchResponse leaks in searchable snapshot test and production code

It's in the title. Fix all the tests like we have everywhere else.
Fix the production use of SearchResponse in
BlobStoreCacheMaintenanceService by properly ref-counting the response
it holds on to.

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

elasticmachine committed:
Merge branch 'main' into fix-leaks-searchable-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into fix-leaks-searchable-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into fix-leaks-searchable-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge branch 'fix-leaks-searchable-snapshots' into ref-count-search-response-itself

original-brownbear committed:
bck

original-brownbear committed:
fix ls

original-brownbear committed:
fix security

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix enrich

original-brownbear committed:
fix monitoring

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix test

original-brownbear committed:
fix test

original-brownbear committed:
fix test

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix test

original-brownbear committed:
fail less hard

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-itself' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
test fixes

original-brownbear committed:
watcher

original-brownbear committed:
fix analytics

original-brownbear committed:
fix rollup

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
closer

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix async

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
fix

original-brownbear committed:
fix more tests

original-brownbear committed:
fixes

original-brownbear committed:
fixes

original-brownbear committed:
fix ds

original-brownbear committed:
fix ds

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
bck

original-brownbear committed:
less diff

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-itself' into ref-count-search-response-itself

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
reduce noise

original-brownbear committed:
reduce noise

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-itself

original-brownbear committed:
start of this

original-brownbear committed:
fix

original-brownbear committed:
Merge branch 'ref-count-search-response-itself' into ref-count-search-response-bytes

original-brownbear committed:
yeay

original-brownbear committed:
cleaner

original-brownbear committed:
cleaner

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
bck

original-brownbear committed:
woerks

original-brownbear committed:
woerks

original-brownbear committed:
woerks

original-brownbear committed:
fixes

original-brownbear committed:
some fixes

original-brownbear committed:
some fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
woerks

original-brownbear committed:
woerks

original-brownbear committed:
we

original-brownbear committed:
eql

original-brownbear committed:
rollup

original-brownbear committed:
rollup

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
fxi to string

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
meh

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
fixes

original-brownbear committed:
fix geoip

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
bck

original-brownbear committed:
LS

original-brownbear committed:
reindex

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
more assert

original-brownbear committed:
percolator

original-brownbear committed:
watcher

original-brownbear committed:
latency simulating

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
fixed?

original-brownbear committed:
bug

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
simple fix?

original-brownbear committed:
unpooled for now

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
another leak

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
revert noise

original-brownbear committed:
revert netty change

original-brownbear committed:
priv

original-brownbear committed:
cleanup

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
weird stuff

original-brownbear committed:
shoter

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
cheaper unpooled

original-brownbear committed:
not needed

original-brownbear committed:
facepalm

original-brownbear committed:
fix pooling

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
rated search hit leak

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
simpler

original-brownbear committed:
massively more efficient

original-brownbear commented:
This is good for review now I believe. I will be adding more detail and some benchmark numbers over the next couple h though.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @original-brownbear, I've created a changelog YAML for you.

original-brownbear committed:
Update docs/changelog/103763.yaml

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear commented on server/src/main/java/org/elasticsearch/action/search/ArraySearchPhaseResults.java:
```diff
@@ -22,7 +22,13 @@
-    private final RefCounted refCounted = LeakTracker.wrap(AbstractRefCounted.of(this::doClose));
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    private final Releasable releasable = LeakTracker.wrap(() -> {
```
Changed plans on this one. Ref-counting was not appropriate here. We want to have this tracked by the context to be release as a last-resort but we also clearly know when this can be released earlier. So I just made it closable with the double-close noop so we can close it wherever.

original-brownbear commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java:
```diff
@@ -217,7 +217,7 @@ public void checkForLeftOverDocuments(Job job, ActionListener<Boolean> listener)
             @Override
             public void onResponse(MultiSearchResponse response) {
-                List<SearchHit> searchHits = new ArrayList<>();
+                List<String> searchHits = new ArrayList<>();
```
This could have been pulled into its own PR, left it here for now to keep things from being blocked again. Just a trivial logical change here to reduce the logical scope of the search hits here.

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-bytes' into ref-count-search-response-bytes

elasticsearchmachine commented:
Hi @original-brownbear, I've updated the changelog YAML for you.

original-brownbear committed:
Update docs/changelog/103763.yaml

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-bytes' into ref-count-search-response-bytes

benwtrent reviewed (commented):
Whew! what a read. This is awesome. 

I eagerly await the performance numbers to see if there is a significant improvement. 

It is frustrating to see how  many non-test things require unPooled :(. But hopefully overtime those things can be updated to ensure we don't copy around bytes needlessly.

benwtrent commented:
@original-brownbear could you post some numbers for a before/after of the benchmarks for this change?

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

original-brownbear commented on modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedSearchHit.java:
```diff
@@ -33,12 +33,12 @@ public class RatedSearchHit implements Writeable, ToXContentObject {
 
     public RatedSearchHit(SearchHit searchHit, OptionalInt rating) {
-        this.searchHit = searchHit;
+        this.searchHit = searchHit.asUnpooled();
```
Yes, this (like pretty much all other spots that use this method) is just a short-cut. Probably easy to fix this in a follow-up :)

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/RandomScoreFunctionIT.java:
```diff
@@ -112,7 +112,7 @@ public void testConsistentHitsWithSameSeed() throws Exception {
                         );
                         final int hitCount = response.getHits().getHits().length;
-                        final SearchHit[] currentHits = response.getHits().getHits();
+                        final SearchHit[] currentHits = response.getHits().asUnpooled().getHits();
```
True but it's more code/change to make that conditional with how things work today. I'd deal with this in a follow-up where I was planning to restrict the explicit use of the array backing a `SearchHits` as much as possible.

original-brownbear commented on server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java:
```diff
@@ -38,11 +38,16 @@ final class FetchSearchPhase extends SearchPhase {
+            resultConsumer,
+            aggregatedDfs,
+            context,
+            (response, queryPhaseResults) -> new ExpandSearchPhase(
+                context,
+                response.hits,
+                () -> new FetchLookupFieldsPhase(context, response, queryPhaseResults)
+            )
+        );
```
Yes exactly :)

original-brownbear commented on server/src/main/java/org/elasticsearch/search/SearchHits.java:
```diff
@@ -63,39 +68,100 @@ public SearchHits(
+            collapseValues,
+            hits.length == 0 ? ALWAYS_REFERENCED : LeakTracker.wrap(new AbstractRefCounted() {
+                @Override
+                protected void closeInternal() {
+                    for (int i = 0; i < hits.length; i++) {
+                        assert hits[i] != null;
+                        hits[i].decRef();
+                        hits[i] = null;
+                    }
+                }
+            })
```
The problem is that we add loads more ref-counting and code if we incref every hit we pass to the constructor of `SearchHits` and dec-ref it where it's constructed. So I went with the pattern of taking ownership here, just like I did for the m-search response for example. I think going forward it would actually make sense to try and move more code to this pattern as an optimisation. Maybe we could even make more stuff closeable instead of ref-counted that way and find more optimisations that way :)

original-brownbear commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -448,6 +497,7 @@ public String getSourceAsString() {
      * The source of the document as a map (can be {@code null}).
      */
     public Map<String, Object> getSourceAsMap() {
+        assert hasReferences();
```
Technically yes, but I went for being strict here. We really shouldn't have code touching released instances ever. Much easier to trust tests if we verify that fact in more rather than fewer spots IMO

original-brownbear commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -463,6 +513,7 @@ public Map<String, Object> getSourceAsMap() {
      * The hit field matching the given field name.
      */
     public DocumentField field(String fieldName) {
+        assert hasReferences();
```
https://github.com/elastic/elasticsearch/pull/103763#discussion_r1453528022 same reasoning here. I'd love to do a follow-up that expands ref counting to the fields in some cases as well.

original-brownbear commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java:
```diff
@@ -52,7 +52,7 @@ public InternalTopHits(
         this.size = size;
         this.topDocs = topDocs;
-        this.searchHits = searchHits;
+        this.searchHits = searchHits.asUnpooled();
```
++

original-brownbear committed:
CR: comments

original-brownbear commented on server/src/main/java/org/elasticsearch/search/SearchHits.java:
```diff
@@ -63,39 +68,100 @@ public SearchHits(
+            sortFields,
+            collapseField,
+            collapseValues,
+            hits.length == 0 ? ALWAYS_REFERENCED : LeakTracker.wrap(new AbstractRefCounted() {
```
True, my reasoning for not adding that check would be that in production usage of this should effectively always work with ref-counted hirs and that condition would force another redundant iteration over the array on us.

original-brownbear commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -653,13 +704,72 @@ public Map<String, Float> getMatchedQueriesAndScores() {
      * @return Inner hits or <code>null</code> if there are none
      */
     public Map<String, SearchHits> getInnerHits() {
+        assert hasReferences();
```
Maybe not right? If we do that here, then we'd have to do it for the source itself before returning it and spots like that also? Seems like consistently asserting ref-counts on fields of ref-counted stuff is going to be a lot of code and doesn't add that much since whoever gets this map is probably going to do something with its contents anyway?

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes

benwtrent reviewed (approved):
Lots to do yet. But this is awesome. I am mostly concerned about multi-node search testing, and CI should handle most of that. We just need to be vigilant.

original-brownbear committed:
Merge remote-tracking branch 'origin/ref-count-search-response-bytes' into ref-count-search-response-bytes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-search-response-bytes
    # Please enter a commit message to explain why this merge is necessary,
    # especially if it merges an updated upstream into a topic branch.
    #
    # Lines starting with '#' will be ignored, and an empty message aborts
    # the commit.

original-brownbear commented:
Wee thanks so much Ben! Merging with Luca's blessing! I'll be on the lookout for leaks :)!

astefan reviewed (commented):
~~LGTM. Thank you, Armin.~~ Wrong PR review.
